generator/mapper.go

```diff
@@ -123,3 +123,72 @@ func getMappedCatalogControlsFromImport(importedCatalog *catalog.Catalog, profil
 	}
 	return newCatalog, nil
 }
+
+func getCatalogForImport(ctx context.Context, i profile.Import, c chan *catalog.Catalog, e chan error) {
+	go func(i profile.Import) {
+		if i.Href == nil {
+			e <- fmt.Errorf("href cannot be nil")
+			return
+		}
+		path, err := GetFilePath(i.Href.String())
+		if err != nil {
+			e <- err
+			return
+		}
+		f, err := os.Open(path)
+		if err != nil {
+			e <- err
+			return
+		}
+		defer f.Close()
+		o, err := oscal.New(f)
+		if err != nil {
+			e <- err
+			return
+		}
+		if o.Catalog != nil {
+			c <- o.Catalog
+			return
+		}
+		for _, p := range o.Profile.Imports {
+			go func(p profile.Import) {
+				getCatalogForImport(ctx, p, c, e)
+			}(p)
+		}
+	}(i)
+}
+
+func AddPartInCatalog(alterations []profile.Alter, c *catalog.Catalog) *catalog.Catalog {
+	for _, x := range alterations {
+		for i, g := range c.Groups {
+			for j, ctrl := range g.Controls {
+				if x.ControlId == "" {
+					continue
+				}
+				if ctrl.Id == x.ControlId {
+					var parts []catalog.Part
+					for _, add := range x.Additions {
+						for _, p := range add.Parts {
+							for _, catalogPart := range ctrl.Parts {
+								var subctrlParts []catalog.Part
+								for k, subctrls := range ctrl.Subcontrols {
+									for _, scp := range subctrls.Parts {
+										if scp.Class == p.Class {
```

Comment:
@anweiss comparing with `class` attribute to append `parts` to the catalog. Is this assumption valid?

Reply:
@minhaj10p so since you're `add`'ing a part, you need to add it regardless of whether or not a `part` with the same `class` already exists in the catalog. However, you should still check to see if a `part` with the same `class` already exists for `id` purposes. In that case, you'll need to append incrementer(s) to the end of the `id` field. So for example:

Original 800-53 catalog part:

```xml
<part class="guidance" id="ca-3.5_gdn">
    <p>Organizations can constrain information system connectivity to external domains (e.g., websites) by employing one of two policies with regard to such connectivity: (i) allow-all, deny by exception, also known as blacklisting (the weaker of the two policies); or (ii) deny-all, allow by exception, also known as whitelisting (the stronger of the two policies). For either policy, organizations determine what exceptions, if any, are acceptable.</p>
</part>
```

FedRAMP High profile `add`:

```xml
<alter subcontrol-id="ca-3.5">
    <add>
        <part class="justification">
             ...
        </part>
        <part class="guidance">
            <p>Guidance: For JAB Authorization, CSPs shall include details of this control in their Architecture Briefing</p>
        </part>
    </add>
</alter>
```

After processing, the final catalog part would look like this:

```xml
<part class="guidance" id="ca-3.5_gdn.1">
    <p>Organizations can constrain information system connectivity to external domains (e.g., websites) by employing one of two policies with regard to such connectivity: (i) allow-all, deny by exception, also known as blacklisting (the weaker of the two policies); or (ii) deny-all, allow by exception, also known as whitelisting (the stronger of the two policies). For either policy, organizations determine what exceptions, if any, are acceptable.</p>
</part>
<part class="guidance" id="ca-3.5_gdn.2">
    <p>Guidance: For JAB Authorization, CSPs shall include details of this control in their Architecture Briefing</p>
</part>
```

Reply:
Thanks for the elaboration! I'll accommodate the changes.